### PR TITLE
Import Fastq Updates

### PR DESCRIPTION
**Pull request by ebelter**
Unzip fastqs
Check fastq line counts 
Verify no reads lost
Remove input fastqs in working dir

### CODE REVIEW

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm [hunk 1]
```diff
@@ -46,16 +60,76 @@ sub execute {
...
+        if ( not -s $unarchived_fastq_path ) {
+            $self->error_message('Unarchived fastq does not exist!');
+            return;
+        }
+        push @new_fastq_paths, $unarchived_fastq_path;
+        unlink $fastq_path;
+    }
+    $self->fastq_paths(\@new_fastq_paths);
+
+    $self->status_message('Unarchive fastqs if necessary...');
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: Should this one say  `done`?

(I remember some time ago we embarked on a project to convert many `status_message`s to `debug_message`s.  I'm wondering if some of these would have been changed had they already been around at that time.)

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm [hunk 2]
```diff
@@ -46,16 +60,76 @@ sub execute {
...
+sub _get_fastq_read_counts {
+    my $self = shift;
+    $self->status_message('Getting fastq read count...');
+
+    my @line_counts;
+    for my $fastq_path ( $self->fastq_paths ) {
+        $self->status_message('Fastq: %s', $fastq_path);
+        open(my $FILE, $fastq_path);
+        1 while <$FILE>;
+        my $line_count = $.;
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: There is `Genome::Sys->line_count` if that seems useful.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm [hunk 3]
```diff
@@ -46,16 +60,76 @@ sub execute {
...
+        $self->fatal_message('Fastq does not have any lines! %s', $fastq_path) if not $line_count > 0;
+        $self->fatal_message('Fastq does not have correct number of lines! %s', $fastq_path) if $line_count % 4 != 0;
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: Most of the other parts of this code are `return`ing on errors (and the call to `_get_fastq_read_counts` checks the return value).  Is it significant that these particular problems are immediately fatal?

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm [hunk 4]
```diff
@@ -46,16 +60,76 @@ sub execute {
...
+    my $read_count = List::Util::sum(@line_counts) / 4;
+    $self->status_message("Fastq read count: $read_count");
+
+    return $self->read_count($read_count);
+}
+
 sub _fastqs_to_bam {
     my $self = shift;
-    $self->debug_message('Run picard fastq to sam...');
+    $self->status_message('Run picard fastq to sam...');
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: Perhaps I spoke too soon about the `debug` vs `status` messages :smile:

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm (hunk 2 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: I I could change Genome::Sys->line_count to match the above. Currently, it relies on a system call  and pattern match. I wanted o avoid a system call and parsing.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm (hunk 3 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: I plan to change all error and return to fatal in the near future.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm (hunk 1 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: I consider these 'status' or INFO level messages. They show the progress of a run. To me DEBUG level is meant for developers to look into internal goings on. What is the proper use case?

**Comment by ebelter** [comment]
@tmooney I made some additional comments - please advise if you can.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm (hunk 2 above)
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: [The perlfaq](http://perldoc.perl.org/perlfaq5.html#How-do-I-count-the-number-of-lines-in-a-file%3f) suggests that shelling out is the fastest option, but maybe one of those chunking algorithms would be suitable?

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm (hunk 3 above)
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/FastqsToBam.pm]: Sounds like a plan!

**Pull request merged**